Increase details of Channels documentation

oxzi opened the pull request:
With a release approaching, the Channels documentation got more information about how the Channel system actually work and how to build your own channels.

The internal/changes/examples JSON files were inlined to ease the readability and updated, as they contained partially outdated information.

For the pkg/plugin part, some API documentation was added as this might get used by external users.

oxzi commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,322 @@
 
-- `method`: The request method to call.
-- `params`: The params for request method.
+- `method`: The request method to call. 
```
:<

oxzi commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,322 @@
+          "en_US": "URL Template"
+        },
+        "help": {
+          "de_DE": "URL, optional als Go-Template über das zu verarbeitende plugin.NotificationRequest.",
+          "en_US": "URL, optionally as a Go template over the current plugin.NotificationRequest."
+        },
+        "required": true,
+        "min": null,
+        "max": null
+      },
+      {
+        "name": "request_body_template",
+        "type": "string",
+        "label": {
+          "de_DE": "Anfragedaten-Template",
+          "en_US": "Request Body Template"
+        },
+        "help": {
+          "de_DE": "Go-Template über das zu verarbeitende plugin.NotificationRequest zum Erzeugen der mitgesendeten Anfragedaten.",
+          "en_US": "Go template applied to the current plugin.NotificationRequest to create an request body."
+        },
+        "default": "{{json .}}",
+        "min": null,
+        "max": null
+      },
+      {
+        "name": "response_status_codes",
+        "type": "string",
+        "label": {
+          "de_DE": "Antwort-Status-Codes",
+          "en_US": "Response Status Codes"
+        },
+        "help": {
+          "de_DE": "Kommaseparierte Liste erwarteter Status-Code der HTTP-Antwort, z.B.: 200,201,202,208,418",
+          "en_US": "Comma separated list of expected HTTP response status code, e.g., 200,201,202,208,418"
+        },
+        "default": "200",
+        "min": null,
+        "max": null
+      }
```
Valid point. It was the output from the webhook channel. I have now reduced it to two options, being used later on.

yhabteab commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,298 @@
+- `id`: The request id.
+
+In case of a present `error` value, the `result` field should be empty.
+Successfully responses without a `result` do only contain the `id` field.
```
This (few other places as well) should be rephrased as [`Successful`](https://dictionary.cambridge.org/dictionary/english/successful) instead.

yhabteab commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,298 @@
+The `config_attrs` field must be an array of JSON objects according to the
+[`ConfigOption` type](https://pkg.go.dev/github.com/icinga/icinga-notifications/pkg/plugin#ConfigOption).
+Those attributes define configuration options for the channel to be set via the `SetConfig` method.
+Furthermore, they are used to build the channel configuration UI in Icinga Notifications Web.
```
I think that `UI` is quite out of place. You might wanted to rephrase it as `... build the UI channel configuration in ...`?

oxzi replied:
I have rephrased the sentence.

yhabteab commented on pkg/plugin/plugin.go:
```diff
@@ -69,57 +69,74 @@ type ConfigOption struct {
 	Address string `json:"address"`
 }
 
+// Object which this NotificationRequest is all about, e.g., a service check.
```
I don't think a `service check` is an object.

oxzi replied:
The text says that the object might be _about_ a service check, not being a service check.

yhabteab replied:
> The text says that the object might be about a service check

That sentence doesn't make sense to me either. An `Event` could be about a service check, but I don't quite understand what an object being about a service check is supposed to mean.

oxzi replied:
Then please take a look at the example `object`: <https://github.com/Icinga/icinga-notifications/blob/17cb89f00b9618c118bc91e2689dd98c65567980/doc/10-Channels.md#example-sendnotification-request>.

I am also open for alternative suggestions.

oxzi replied:
Or was your point that it does not refer to a service check but more to a service object?

_Btw, I find the name `object` hard to grasp, especially as both in Icinga 2 and OOP lingo everything is a object._

martialblog commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,298 @@
 - `method`: The request method to call.
-- `params`: The params for request method.
+- `params`: Optional params for the request method.
 - `id`: Unsigned int value. Required to assign the response to its request as responses can be sent out of order.
```
Hi, 

the field description could be normalized:

`fieldname: description of what it does. optional/madatory. data type`

Regards

yhabteab replied:
Yes, maybe as a markdown table, like [Icinga 2](https://icinga.com/docs/icinga-2/latest/doc/12-icinga2-api/#creating-config-objects) does.

oxzi replied:
Thanks., I liked this idea. I tried to unify all fields to:
> name: Short description. Type.

oxzi replied:
I switched to a table, as you suggested @yhabteab. This has the only downside that diffs altering the column width span over the whole table. But, it looks nicer :)

oxzi replied:
@yhabteab: I tried to mimic Icinga 2's layout.

martialblog commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,298 @@
+
+Each request must be answered by the channel with a response JSON object of the following fields:
+
+- `result`: The result as JSON format. Omitted when the method does not return a value, i.e., for setter calls, or an
```
Again, normalization could be useful:

`fieldname: description of what it does. optional/madatory. data type`

martialblog commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,298 @@
+- `error`: The error message. Omitted when no error has occurred.
+- `id`: The request id.
+
+In case of a present `error` value, the `result` field should be empty.
```
Does "empty" mean `null` or the field can be omitted. Or does it mean empty string?

martialblog commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,298 @@
+#### SetConfig
+
+The `SetConfig` method configures the channel and
+should be called at least once before sending the first notifications to initialize the channel plugin.
```
"should be called" by whom? The channel plugin?

oxzi replied:
Good point. In the eyes of a channel developer, this is straight out confusing.

martialblog commented on doc/10-Channels.md:
````diff
@@ -1,88 +1,298 @@
+}
+```
+
+The response might carry an error if the configuration is invalid.
````
What does "might" mean here? Optional? Recommended?

Might I suggest using the key words from RFC2119?

oxzi replied:
`s/might/MAY/g` :^)

I rewrote some sections to remove inaccuracies, use some RFC 2119 language (just w/o the caps), and to address a channel developer.

martialblog commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,298 @@
+[`NotificationRequest`](https://pkg.go.dev/github.com/icinga/icinga-notifications/pkg/plugin#NotificationRequest)
+is expected.
+
+The response might carry an error due to channel-specific reasons.
```
Maybe add an example of what an error response

oxzi commented:
I reworked some parts, especially to mention changes from #206. Furthermore, I described how the channel configuration through the two RPC methods works and how a channel plugin is expected to behave. As the documentation links to the Go API doc, I added more API docs for the fields.

Please take another look.

yhabteab commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,355 @@
 
+If the channel plugin successfully configured itself, a response without an `result` must be returned.
+Otherwise, if the channel decides that the provided configuration is faulty, an `error` must be returned.
+This may happen if, i.e., an invalid configuration value was given.
```
I think you don't want to use [`i.e.`](https://dictionary.cambridge.org/dictionary/english/ie) here, but `for instance/example` instead.

oxzi replied:
There will be a day when I finally learn how to correctly use _e.g._ and _i.e._

oxzi commented on doc/10-Channels.md:
```diff
@@ -1,88 +1,317 @@
-- `id`: Unsigned int value. Required to assign the response to its request as responses can be sent out of order.
+### Response
+
+Each request must be answered by the channel with a response JSON object of the following fields:
```
I removed the "response" but kept JSON object to exactly specify what we are talking about. I don't want to rely on too much implied information.

yhabteab reviewed (approved):
LFTM!

julianbrost commented:
@martialblog Do you want to have another look as well, i.e. should I wait with merging this a bit?

martialblog commented:
@julianbrost Thanks, I was away these last days. If I find something I'll just open a PR. 

Looking forward to v0.1.0
